src/iftraf.c

```diff
@@ -0,0 +1,1051 @@
+static void inline iftraf_iftlb_add(struct iftraf_pkt *param)
+{
+    uint32_t hash;
+    struct iftraf_entry *entry = NULL;
+	
+    hash = iftraf_byif_hashkey(param->af, &param->saddr, param->devid);
+	
+    if (iftraf_ifentry_get(hash, param, &entry) == HASH_STATUS_KEY_NOT_FOUND) {
+
+        entry = rte_zmalloc(NULL, sizeof(struct iftraf_entry), RTE_CACHE_LINE_SIZE);
+        if (entry == NULL) {
+            RTE_LOG(ERR, IFTRAF,
+                "%s: no memory\n", __func__);
+            return;
+        }
+		
+        memset(entry, 0, sizeof(struct iftraf_entry));
+        entry->af = param->af;
+        entry->cid = param->cid;
+        entry->devid = param->devid;
+        entry->proto = param->proto;
+        iftraf_addr_cpy(param->af, &entry->saddr, &param->saddr);
+        iftraf_addr_cpy(param->af, &entry->daddr, &param->daddr);
+        entry->sport = param->src_port;
+        entry->dport = param->dst_port;
+
+        list_add(&entry->list, &iftraf_iftbl[hash]);				  
+    }
+	
+    if (param->af == AF_INET) {
+        RTE_LOG(DEBUG, IFTRAF,
+            "%s:[v4] dequeue iftraf_ring[cid:%d, proto:%u, src:%08X, dst:%08X, sp:%u, dp:%u, len:%u]\n",
+            __func__, entry->cid, entry->proto, entry->saddr.in.s_addr, 
+            entry->daddr.in.s_addr, ntohs(entry->sport), ntohs(entry->dport), param->pkt_len);
+    } else {
+        RTE_LOG(DEBUG, IFTRAF,
+            "%s:[v6] dequeue iftraf_ring[cid:%d, dir:%d, proto:%u, src:%08X %08X %08X %08X, dst:%08X %08X %08X %08X, sp:%u, dp:%u, len:%u]\n",
+            __func__, entry->cid, param->dir, entry->proto, 
+            entry->saddr.in6.s6_addr32[0], entry->saddr.in6.s6_addr32[1],
+            entry->saddr.in6.s6_addr32[2], entry->saddr.in6.s6_addr32[3],
+            entry->daddr.in6.s6_addr32[0],entry->daddr.in6.s6_addr32[1],
+            entry->daddr.in6.s6_addr32[2],entry->daddr.in6.s6_addr32[3],
+            ntohs(entry->sport), ntohs(entry->dport), param->pkt_len);
+    }
+
+    /* Update record */
+    entry->last_write = history_pos;
+    if (param->dir == IFTRAF_PKT_DIR_IN) {
+        entry->recv[history_pos] += param->pkt_len;
+        entry->total_recv += param->pkt_len;
+        RTE_LOG(DEBUG, IFTRAF,
+        "%s: history_pos: %d, recv : %u, total_recv : %u\n", __func__, history_pos, entry->recv[history_pos], entry->total_recv);
+
+    } else {
+        entry->sent[history_pos] += param->pkt_len;
+        entry->total_sent += param->pkt_len;
+
+        RTE_LOG(DEBUG, IFTRAF,
+            "%s: history_pos: %d: sent : %u, total_sent: %u\n", __func__, history_pos, entry->sent[history_pos], entry->total_sent);
+    }
+}
+
+
+void iftraf_process_ring(void)
+{    
+    int i;
+    uint16_t nb_rb;
+    lcoreid_t cid;
+    struct iftraf_pkt *param;
+    struct iftraf_pkt *params[NETIF_MAX_PKT_BURST];
+	
+
+    if (likely(iftraf_disable)) {		
+        return;
+    }	
+
+    ticket++;
+    if(ticket % TIME_INTERVAL == 0) {
+        history_rotate();
+        ticket = 0;
+    }

```

Comment:
1. "history_rotate" is called less frequently if traffic is heavy, so we cannot get the topN traffic in Bps or bps because no time info is recorded.  If we remember the timestamp or cpu ticks every time we call "history_rotate", can we figure out a solution to acquire the approximation of the topN traffic per second?
2. How long does "history_rotate" take when traffic is heavy? Would it starve other jobs in this CPU?